core/app/models/spree/promotion_handler/promotion_batch_duplicator.rb

```diff
@@ -0,0 +1,37 @@
+module Spree
+  module PromotionHandler
+    class PromotionBatchDuplicator < Spree::PromotionDuplicatorCore
+      def initialize(promotion, promotion_batch_id, random_string: nil, code: nil)
+        @promotion = promotion
+        @promotion_batch_id = promotion_batch_id
+        @random_string = random_string || generate_random_string(4)
+        @code = code
+      end
+
+      def duplicate
+        @new_promotion = @promotion.dup
+        @new_promotion.usage_limit = 1
+        @new_promotion.promotion_batch_id = @promotion_batch_id
+        @new_promotion.path = "#{@promotion.path}_#{@random_string}"
+        code_assignment
+        @new_promotion.stores = @promotion.stores
+
+        ActiveRecord::Base.transaction do
+          @new_promotion.save
+          copy_rules
+          copy_actions
+        end
+
+        @new_promotion
+      end
+
+      private
+
+      def code_assignment
+        return @new_promotion.generate_code=(true) unless @code
+
+        @new_promotion.code = @code
+      end
```

Comment:
Hm, the fact that the duplicated promotion is saved as an instance variable is weird IMO. It introduces some unnecessary state to track. Wouldn't a simple variable be easier here? It's scope would be much narrower.

Reply:
I know that it's used by the `Core` class but maybe there are better ways of structuring this? For example just passing the promotion to `copy_rules` and `copy_actions`. Then the "contract" is simpler IMO, makes the entire thing easier to follow because all the variables are local.

Reply:
Changed.